single-line-diagram/single-line-diagram-core/src/main/java/com/powsybl/sld/layout/MatrixZoneLayout.java

```diff
@@ -0,0 +1,127 @@
+ * SPDX-License-Identifier: MPL-2.0
+ */
+package com.powsybl.sld.layout;
+
+import com.powsybl.sld.layout.pathfinding.*;
+import com.powsybl.sld.layout.zonebygrid.*;
+import com.powsybl.sld.model.coordinate.*;
+import com.powsybl.sld.model.coordinate.Point;
+import com.powsybl.sld.model.graphs.*;
+import com.powsybl.sld.model.nodes.Node;
+import org.jgrapht.alg.util.*;
+
+import java.util.*;
+
+/**
+ * @author Thomas Adam <tadam at silicom.fr>
+ */
+public class MatrixZoneLayout extends AbstractZoneLayout {
+    private final MatrixZoneLayoutModel model;
+    private final String[][] matrix;
+
+    private final PathFinder pathFinder;
+
+    protected MatrixZoneLayout(ZoneGraph graph, String[][] matrix, SubstationLayoutFactory sLayoutFactory, VoltageLevelLayoutFactory vLayoutFactory) {
+        super(graph, sLayoutFactory, vLayoutFactory);
+        this.model = new MatrixZoneLayoutModel(90);
+        this.matrix = matrix;
+        this.pathFinder = new PathFinderFactory().createDijkstra();
+    }
+
+    /**
+     * Calculate relative coordinate of substations in the zone
+     */
+    @Override
+    protected void calculateCoordSubstations(LayoutParameters layoutParameters) {
+        for (int row = 0; row < matrix.length; row++) {
+            for (int col = 0; col < matrix[row].length; col++) {
+                String id = matrix[row][col];
+                SubstationGraph graph = getGraph().getSubstationGraph(id);
+                if (graph != null) {
+                    // Display substations
+                    layoutBySubstation.get(graph).run(layoutParameters);
+                }
+                model.addGraph(graph, col, row);
+            }
+        }
+        // Height by rows
+        int maxHeightRow = model.getMatrixCellHeight();
+        // Width by col
+        int maxWidthCol = model.getMatrixCellWidth();
+        // Snakeline hallway
+        int hallway = model.getSnakelineHallwayWidth();
+        // Padding
+        LayoutParameters.Padding diagramPadding = layoutParameters.getDiagramPadding();
+        // Zone size
+        double zoneWidth = 0.0;
+        double zoneHeight = diagramPadding.getTop() + hallway;
+        // Move each substation into its matrix position
+        for (int row = 0; row < matrix.length; row++) {
+            zoneWidth = hallway;
+            int hallwayRow = (row + 1) * hallway;
+            for (int col = 0; col < matrix[row].length; col++) {
+                String id = matrix[row][col];
+                SubstationGraph graph = getGraph().getSubstationGraph(id);
+                int hallwayCol = (col + 1) * hallway;
+                if (graph != null) {
+                    double dx = col * maxWidthCol + hallwayCol + diagramPadding.getLeft();
+                    double dy = row * maxHeightRow + hallwayRow + diagramPadding.getTop();
+                    move(graph, dx, dy);
+                }
+                zoneWidth += maxWidthCol + hallway;
```

Comment:
I think `zoneWidth` and `zoneHeight` can be computed at the beginning. Is there a need to compute them step by step in the loops?

Reply:
In deed